### PR TITLE
refactor(test): support custom writer in PrettyTestReporter

### PR DESCRIPTION
**Pull request by nayeemrmn**
Extracted from #20778.

### CODE REVIEW

File: cli/tools/test/reporters/pretty.rs [hunk 1]
```diff
@@ -139,19 +159,25 @@ impl TestReporter for PrettyTestReporter {
   fn report_plan(&mut self, plan: &TestPlan) {
     self.summary.total += plan.total;
     self.summary.filtered_out += plan.filtered_out;
+    if self.repl {
+      writeln!(&mut self.writer).unwrap();
+      return;
+    }
```
Comment (bartlomieju) [inline cli/tools/test/reporters/pretty.rs:165]: Why this single line is needed here?

File: cli/tools/test/reporters/pretty.rs [hunk 2]
```diff
@@ -331,7 +368,10 @@ impl TestReporter for PrettyTestReporter {
     _tests: &IndexMap<usize, TestDescription>,
     _test_steps: &IndexMap<usize, TestStepDescription>,
   ) {
-    common::report_summary(&self.cwd, &self.summary, elapsed);
+    if self.repl && self.summary.total == 0 {
+      return;
+    }
```
Comment (bartlomieju) [inline cli/tools/test/reporters/pretty.rs:373]: Are we gonna try to run tests on every expression evaluation in the REPL?

File: cli/tools/test/reporters/pretty.rs (hunk 1 above)
Comment (nayeemrmn) [reply cli/tools/test/reporters/pretty.rs]: > Why this single line is needed here?

Removed, don't remember why I added it.

File: cli/tools/test/reporters/pretty.rs (hunk 2 above)
Comment (nayeemrmn) [reply cli/tools/test/reporters/pretty.rs]: > Are we gonna try to run tests on every expression evaluation in the REPL?

It doesn't do that much work but I guess it adds two extra `.await`ed calls on every eval. I'll explicitly check for tests in the repl instead.

**Review by bartlomieju — APPROVED** [review]

**Pull request merged**